controllers/volumereplicationgroup_controller.go

```diff
@@ -984,13 +986,18 @@ func (v *VRGInstance) reconcileAsPrimary() ctrl.Result {
 	result := ctrl.Result{}
 	result.Requeue = v.reconcileVolSyncAsPrimary(&finalSyncPrepared.volSync)
 	v.reconcileVolRepsAsPrimary(&result.Requeue)
-	v.kubeObjectsProtectPrimary(&result, s3StoreAccessors)
+	v.kubeObjectsReconcilePrimary(&result, s3StoreAccessors)
 	v.vrgObjectProtect(&result, s3StoreAccessors)
 
 	if vrg.Spec.PrepareForFinalSync {
 		vrg.Status.PrepareForFinalSyncComplete = finalSyncPrepared.volSync
 	}
 
+	v.statusConditionSetFirstFalseOrLastTrue(VRGConditionTypeClusterDataReady,
```

Comment:
Overall, I have concerns about the flow. I'm usually not a fan of complex function arguments used for function calls. It really adds a lot of complexity. And it is becoming a pattern now.  It was difficult to understand the changes. 

Anyway, based on Travis' description, it seems that the main modification involves moving the kube object restore to a later stage, specifically after ensuring that the volume is available via `VR` and ready through the `DataReady` condition, which seems logical given the context. In this line, however, the `DataReady` status condition will never be updated to true as long as the `kubeObjectsRecovered` is false. The status won't be true because you also changed line 709, which made the `DataReady` condition values temporary values until they are saved in `etcd`. But they will never be saved due to line 711.

Again, I admit it, I don't understand fully the change without spending too much time reviewing this.

Reply:
> In this line, however, the `DataReady` status condition will never be updated to true as long as the `kubeObjectsRecovered` is false. The status won't be true because you also changed line 709, which made the `DataReady` condition values temporary values until they are saved in `etcd`. But they will never be saved due to line 711.

I presume you meant to say `ClusterDataReady` not `DataReady` because the lines referenced set `ClusterDataReady`.  However, the main purpose of this patch is to wait for `DataReady` to be set before recovering the Kube resources for the case where a hook needs to be run in a pod and the pod has volumes that need to be ready.  But you're right that another reconcile will always be required since `DataReady` is set later.  To avoid that, the setting of `DataReady` could be moved up or the recovery of the Kube objects could be moved down or even out to another controller...something I'm considering for the multi-namespaced application.